Azure/azure-sdk-for-net
# EventGrid blob trigger support

Current Blob trigger logic relies on [pulling](https://github.com/Azure/azure-webjobs-sdk/pull/2438) azure storage diagnostic logs. This approach gives us high latencies, bad performance and fragile behavior overall. Using EventGrid subscription is much more reliable approach and it will give approximately 2-3 times better performance.

The overall flow for EventGrid Blob trigger: 
1. A blob uploaded to azure storage.
2. EventGrid subscription detects the blob change and sends a HTTP request to the blob extension HTTP endpoint.
3. Blob extension receives the HTTP request, parses it and put a new message to the internal storage queue with the blob information and function name. Regular Blob trigger uses the same internal queue. 
4. Internal storage queue listener reads the new queue message, creates CloudBlob and executes the function code. 

Steps to test locally:
1. Create a function app with blob trigger in VS.
2. Add new nuget source: https://www.myget.org/F/azure-appservice/api/v3/index.json
3. Replace the storage extension package with `Azure.WebJobs.Extensions.Storage.Blobs.5.0.0-alpha.20201120.1`
4. Add `"UseEventGrid = true`" to trigger definition
`public static void Run([BlobTrigger("samples-workitems/{name}", UseEventGrid = true)]Stream myBlob, string name, ILogger log)`
5. Start ngrok http -host-header=localhost 7071
https://docs.microsoft.com/en-us/azure/azure-functions/functions-debug-event-grid-trigger-local
6. Start the function app VS project.
7. Open portal and create an EventGrid subscription with WebHook endpoint:
https://83f2d199d6f4.ngrok.io/runtime/webhooks/blobs?functionName=Function1
Note: EventGrid subscription fires on all containers/blobs in the storage account. Maybe you want to add some EventGrid subscription filters.
Note: If you deploy the function app to azure the EventGrid subscription endpoint will be:
https://[functionAppName].azurewebsites.net/runtime/webhooks/blobs?functionName=Function1&code=[blobExtensionKey]
BlobExtensionKey is generated automatically and available on the azure portal.
8. Add a blob to blob storage and verify the trigger is fired.

The plan is to provide the EventGrid Blob trigger in beta as is, write documentation and collect customers feedback.
